localstack/services/s3/s3_listener.py

```diff
@@ -193,6 +195,94 @@ def strip_chunk_signatures(data):
     return data_new
 
 
+def _iter_multipart_parts(some_bytes, boundary):
+    ''' Generate a stream of dicts and bytes for each message part.
+
+        Content-Disposition is used as a header for a multipart body:
+        https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition
+    '''
+    parser = email.parser.BytesHeaderParser()
+
+    while True:
+        try:
+            part, some_bytes = some_bytes.split(boundary, 1)
+        except ValueError:
+            # Ran off the end, stop.
+            break
+
+        if b'\r\n\r\n' not in part:
+            # Real parts have headers and a value separated by '\r\n'.
+            continue
+
+        part_head, _ = part.split(b'\r\n\r\n', 1)
+        head_parsed = parser.parsebytes(part_head.lstrip(b'\r\n'))
+
+        if 'Content-Disposition' in head_parsed:
+            _, params = cgi.parse_header(head_parsed['Content-Disposition'])
+            yield params, part
+
+
+def expand_multipart_filename(data, headers):
+    ''' Replace instance of '${filename}' in key with given file name.
+
+        Data is given as multipart form submission bytes, and file name is
+        replace according to Amazon S3 documentation for Post uploads:
+        http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html
+    '''
+    _, params = cgi.parse_header(headers.get('Content-Type'))
+    boundary = params['boundary'].encode('ascii')
+    bytes = data.encode('utf8') if (type(data) is str) else data
```

Comment:
We can reuse the `to_bytes` method from `common.py` here. Also, `bytes` is a reserved word in Python, let's better avoid using it.
```
data_bytes = to_bytes(data)
```